### PR TITLE
#1322 Make _INFO files validation a configurable option

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
3 types of validation + verbose errors
Closes [1322](https://github.com/AbsaOSS/enceladus/issues/1322) fixes [1317](https://github.com/AbsaOSS/enceladus/issues/1317)

### CODE REVIEW

File: spark-jobs/src/main/resources/application.conf.template [hunk 1]
```diff
@@ -42,6 +42,9 @@ conformance.catalyst.workaround=true
 # Automatically delete standardized data folder after successful run of a Conformance job
 conformance.autoclean.standardized.hdfs.folder=false
 
+#enforce info file validation
```
Comment (Zejnilovic) [inline spark-jobs/src/main/resources/application.conf.template:45]: add all options here please

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala [hunk 2]
```diff
@@ -26,28 +28,43 @@ object ControlInfoValidation {
...
+        validateFields(checkpointRawRecordCount, checkpointSourceRecordCount)
+    } catch {
+      case ex: ValidationException => controlInfoValidation match {
+        case "strict" => throw ex
+        case "warning" => log.warn(ex.msg)
+        case _ =>
+      }
+    } finally {
+      checkpointRawRecordCount
+        .foreach(e => Atum.setAdditionalInfo(s"${rawFieldName}_record_count" -> e.toString))
```
Comment (Zejnilovic) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala:49]: Not sure. Why is this moved to additional info?

Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala:49]: These fields in additional info help Menas Monitoring and [possibly] SPOT to parse _INFO files. Quite convenient.

File: spark-jobs/src/test/resources/application.conf [hunk 3]
```diff
@@ -28,5 +28,8 @@ conformance.mapping.rule.broadcast=never
 # Maximum size (in MB) of a mapping table to use the efficient broadcasting mapping rule strategy
 conformance.mapping.rule.max.broadcast.size.mb=0
 
+#enforce info file validation
```
Comment (yruslan) [inline spark-jobs/src/test/resources/application.conf:31]: in addition to @Zejnilovic suggestion:
```suggestion
# Enforce info file validation
```

File: spark-jobs/src/main/resources/application.conf.template [hunk 4]
```diff
@@ -42,6 +42,10 @@ conformance.catalyst.workaround=true
 # Automatically delete standardized data folder after successful run of a Conformance job
 conformance.autoclean.standardized.hdfs.folder=false
 
+# Enforce info file validation
+# Can be one of: strict, warning, off (anything else will be treated as off)
+control.info.validation=warning
```
Comment (dk1844) [inline spark-jobs/src/main/resources/application.conf.template:47]:
```suggestion
standardized.control.info.validation=warning
```

```suggestion
standardization.control.info.validation=warning
```

If this is for standardization only, maybe the key could reflect that somehow?

Comment (AdrianOlosutean) [reply spark-jobs/src/main/resources/application.conf.template:47]: Actually it should have applied for conformance too, good point. I will use it in conformance too

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 5]
```diff
@@ -52,6 +52,7 @@ object StandardizationJob {
   private val log = LoggerFactory.getLogger(this.getClass)
   private val conf = ConfigFactory.load()
   private val menasBaseUrls = MenasConnectionStringParser.parse(conf.getString("menas.rest.uri"))
+  private val controlInfoValidation = conf.getString("control.info.validation")
```
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:55]: Maybe this would deserve to get extracted somewhere (perhaps into `ControlInfoValidation`) -- given the fact that it is shared with Conformance?

**Comment by yruslan** [comment]
Looks good, but need to resolve conflicts before the merge.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala [hunk 6]
```diff
@@ -25,29 +28,46 @@ import scala.util.{Failure, Success, Try}
...
+
+      checkpointSourceRecordCount
+        .foreach(e => Atum.setAdditionalInfo(s"${sourceFieldName}_record_count" -> e.toString))
+    }
+  }
+
+  def validateFields(checkpointRawRecordCount: Try[Long], checkpointSourceRecordCount: Try[Long]) {
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala:59]: Would make it a function returning `Try[Unit]`. Also would make it `private` I don't think this needs such a granular level level of unit-testing.
(I believe it's consistent and better to read if one sticks within one class/flo to either use `class Try` xor `try...catch` block.)

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala [hunk 7]
```diff
@@ -25,29 +28,46 @@ import scala.util.{Failure, Success, Try}
...
    * Adds metadata about the number of records in raw and source data by checking Atum's checkpoints first.
    */
-  def addRawAndSourceRecordCountsToMetadata(): Unit = {
-    val fields = getFieldCounts(Atum.getControlMeasure.checkpoints)
-    Atum.setAdditionalInfo(s"${rawFieldName}_record_count" -> fields.rawCount.toString)
-    Atum.setAdditionalInfo(s"${sourceFieldName}_record_count" -> fields.sourceCount.toString)
-  }
-
-  def getFieldCounts(checkpoints: Seq[Checkpoint]): FieldCounts = {
+  def addRawAndSourceRecordCountsToMetadata(conf: Config, log: Logger): Unit = {
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala:36]: So should writint to the INFO file also happen in the job?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 8]
```diff
@@ -333,6 +332,16 @@ object StandardizationJob {
     standardizedDF.writeInfoFile(pathCfg.outputPath)
   }
 
+  private def handleControlInfoValidation(): Unit = {
+    ControlInfoValidation.addRawAndSourceRecordCountsToMetadata() match {
+      case Failure(ex: ValidationException) => conf.getString("control.info.validation") match {
+        case "strict" => throw ex
+        case "warning" => log.warn(ex.msg)
+        case _ =>
```
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:340]: This works as described, however, I wonder if it truly desired that invalid/misspelled value results in no action. I am personally a fan of strictly checking that only allowed enum-like values are used and verbosely failing otherwise.

Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:340]: I was also thinking about this one. What do the others think? @benedeki @yruslan @Zejnilovic

Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:340]: I'd prefer strict checking as well. Otherwise, a typo (i.e. `warming`) could lead to unexpected behavior that is not easy to spot.

Comment (Zejnilovic) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:340]: options are strict, warning and none? I would just check for `warning` and `none`, everything else is a `throw`.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala [hunk 9]
```diff
@@ -318,6 +320,19 @@ object DynamicConformanceJob {
     }
   }
 
+  private def handleControlInfoValidation(): Unit = {
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala:323]: You have this method written in two places. Can we only one of them in a single place? Maybe inside `ControlInfoValidation`?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala [hunk 10]
```diff
@@ -318,6 +320,19 @@ object DynamicConformanceJob {
...
+    ControlInfoValidation.addRawAndSourceRecordCountsToMetadata() match {
+      case Failure(ex: za.co.absa.enceladus.utils.validation.ValidationException) =>
+        val confEntry = "control.info.validation"
+        conf.getString(confEntry) match {
+          case "strict" => throw ex
+          case "warning" => log.warn(ex.msg)
+          case "none" =>
+          case _ => throw new ValidationException(s"Invalid $confEntry value")
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala:331]: `ValidationException` is more about validating user input, not a config. Maybe `RuntimeException`?

File: spark-jobs/src/main/resources/application.conf.template [hunk 11]
```diff
@@ -46,6 +46,10 @@ conformance.catalyst.workaround=true
 # Automatically delete standardized data folder after successful run of a Conformance job
 conformance.autoclean.standardized.hdfs.folder=false
 
+# Enforce info file validation
+# Can be one of: strict, warning, off
```
Comment (yruslan) [inline spark-jobs/src/main/resources/application.conf.template:50]:
```suggestion
# Can be one of: strict, warning, none
```

File: spark-jobs/src/test/resources/application.conf [hunk 12]
```diff
@@ -28,5 +28,9 @@ conformance.mapping.rule.broadcast=never
 # Maximum size (in MB) of a mapping table to use the efficient broadcasting mapping rule strategy
 conformance.mapping.rule.max.broadcast.size.mb=0
 
+# Enforce info file validation
+# Can be one of: strict, warning, off
```
Comment (yruslan) [inline spark-jobs/src/test/resources/application.conf:32]:
```suggestion
# Can be one of: strict, warning, none
```

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala (hunk 9 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala:323]: Talking with @benedeki and we thought we could leave this for when the refactoring of standardization and conformance

Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/DynamicConformanceJob.scala:323]: Yes, we discussed this with Adrian.  I commented that the  original `ControlInfoValidation` was starting doding little too much. And from the options what and how to etxract. we though this to be the way to go. Also considered to put this logic to one object, not to duplicate code, but it felt like a too empty shell.
